Sources/NIO/ControlMessage.swift

```diff
@@ -0,0 +1,245 @@
+struct UnsafeControlMessage {
+    var level: CInt
+    var type: CInt
+    var data: UnsafeRawBufferPointer?
+}
+
+/// Collection representation of `cmsghdr` structures and associated data from `recvmsg`
+/// Unsafe as captures pointers held in msghdr structure which must not escape scope of validity.
+struct UnsafeControlMessageCollection {
+    private var messageHeader: msghdr
+
+    init(messageHeader: msghdr) {
+        self.messageHeader = messageHeader
+    }
+}
+
+// Add the `Collection` functionality to UnsafeControlMessageCollection.
+extension UnsafeControlMessageCollection: Collection {
+    typealias Element = UnsafeControlMessage
+    
+    struct Index: Equatable, Comparable {
+        fileprivate var cmsgPointer: UnsafeMutablePointer<cmsghdr>?
+        
+        static func < (lhs: UnsafeControlMessageCollection.Index,
+                       rhs: UnsafeControlMessageCollection.Index) -> Bool {
+            // nil is high, as that's the end of the collection.
+            switch (lhs.cmsgPointer, rhs.cmsgPointer) {
+            case (.some(let lhs), .some(let rhs)):
+                return lhs < rhs
+            case (.some, .none):
+                return true
+            case (.none, .some), (.none, .none):
+                return false
+            }
+        }
+        
+        fileprivate init(cmsgPointer: UnsafeMutablePointer<cmsghdr>?) {
+            self.cmsgPointer = cmsgPointer
+        }
+    }
+    
+    var startIndex: Index {
+        var messageHeader = self.messageHeader
+        return withUnsafePointer(to: &messageHeader) { messageHeaderPtr in
+            let firstCMsg = Posix.cmsgFirstHeader(inside: messageHeaderPtr)
+            return Index(cmsgPointer: firstCMsg)
+        }
+    }
+    
+    var endIndex: Index { return Index(cmsgPointer: nil) }
+    
+    func index(after: Index) -> Index {
+        precondition(after.cmsgPointer != nil)
+        var msgHdr = messageHeader
+        return withUnsafeMutablePointer(to: &msgHdr) { messageHeaderPtr in
+            return Index(cmsgPointer: Posix.cmsgNextHeader(inside: messageHeaderPtr,
+                                                           after: after.cmsgPointer!))
+        }
+    }
+    
+    public subscript(position: Index) -> Element {
+        let cmsg = position.cmsgPointer!
+        return UnsafeControlMessage(level: cmsg.pointee.cmsg_level,
+                                    type: cmsg.pointee.cmsg_type,
+                                    data: Posix.cmsgData(for: cmsg))
+    }
+}
+
+/// Extract information from a collection of control messages.
+struct ControlMessageParser {
+    var ecnValue: NIOExplicitCongestionNotificationState = .transportNotCapable // Default
+    
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    private static let ipv4TosType = IP_RECVTOS
+    #else
+    private static let ipv4TosType = IP_TOS    // Linux
+    #endif
+
+    static func readCInt(data: UnsafeRawBufferPointer) -> CInt {
```

Comment:
Prefix with an underscore to indicate this is not supposed to be part of the public API of this type.